File: scripts/load_test.py
#!/usr/bin/env python3
"""Small standard-library load test for ferryllm.

This is intended for local mock-upstream testing, not for real paid providers.
"""

from __future__ import annotations

import argparse
import concurrent.futures
import json
import statistics
import time
import urllib.error
import urllib.request


def request_once(url: str, timeout: float) -> tuple[int, float]:
    body = json.dumps(
        {
            "model": "claude-load-test",
            "max_tokens": 16,
            "messages": [{"role": "user", "content": "ping"}],
        }
    ).encode("utf-8")
    request = urllib.request.Request(
        url,
        data=body,
        headers={"content-type": "application/json", "authorization": "Bearer load-test"},
        method="POST",
    )
    started = time.perf_counter()
    try:
        with urllib.request.urlopen(request, timeout=timeout) as response:
            response.read()
            status = response.status
    except urllib.error.HTTPError as exc:
        exc.read()
        status = exc.code
    except Exception:
        status = 0
    return status, (time.perf_counter() - started) * 1000.0


def percentile(values: list[float], pct: float) -> float:
    if not values:
        return 0.0
    ordered = sorted(values)
    index = min(len(ordered) - 1, int(round((pct / 100.0) * (len(ordered) - 1))))
    return ordered[index]


def main() -> int:
    parser = argparse.ArgumentParser(description="Load test ferryllm with a mock upstream")
    parser.add_argument("--url", default="http://127.0.0.1:3000/v1/messages")
    parser.add_argument("--requests", type=int, default=1000)
    parser.add_argument("--concurrency", type=int, default=32)
    parser.add_argument("--timeout", type=float, default=10.0)
    args = parser.parse_args()

    started = time.perf_counter()
    statuses: dict[int, int] = {}
    latencies: list[float] = []

    with concurrent.futures.ThreadPoolExecutor(max_workers=args.concurrency) as pool:
        futures = [pool.submit(request_once, args.url, args.timeout) for _ in range(args.requests)]
        for future in concurrent.futures.as_completed(futures):
            status, latency_ms = future.result()
            statuses[status] = statuses.get(status, 0) + 1
            latencies.append(latency_ms)

    elapsed = time.perf_counter() - started
    ok = statuses.get(200, 0)
    errors = args.requests - ok
    print(json.dumps(
        {
            "requests": args.requests,
            "concurrency": args.concurrency,
            "elapsed_seconds": round(elapsed, 3),
            "requests_per_second": round(args.requests / elapsed, 2) if elapsed else 0,
            "ok": ok,
            "errors": errors,
            "statuses": statuses,
            "latency_ms": {
                "min": round(min(latencies), 2) if latencies else 0,
                "mean": round(statistics.mean(latencies), 2) if latencies else 0,
                "p50": round(percentile(latencies, 50), 2),
                "p95": round(percentile(latencies, 95), 2),
                "p99": round(percentile(latencies, 99), 2),
                "max": round(max(latencies), 2) if latencies else 0,
            },
        },
        indent=2,
        sort_keys=True,
    ))
    return 0 if errors == 0 else 1


if __name__ == "__main__":
    raise SystemExit(main())

File: examples/load_test.rs
//! Async Rust load tester for ferryllm.
//!
//! Example:
//!   cargo run --release --example load_test --features http -- --requests 10000 --concurrency 512

use std::collections::BTreeMap;
use std::sync::Arc;
use std::time::{Duration, Instant};

use reqwest::Client;
use serde_json::json;
use tokio::sync::Semaphore;

#[derive(Clone)]
struct Args {
    url: String,
    requests: usize,
    concurrency: usize,
    timeout_secs: u64,
}

#[derive(Debug)]
struct ResultSample {
    status: u16,
    latency_ms: f64,
}

#[tokio::main]
async fn main() {
    let args = parse_args();
    let client = Client::builder()
        .pool_max_idle_per_host(args.concurrency)
        .timeout(Duration::from_secs(args.timeout_secs))
        .build()
        .expect("build client");
    let semaphore = Arc::new(Semaphore::new(args.concurrency));
    let started = Instant::now();
    let mut tasks = Vec::with_capacity(args.requests);

    for _ in 0..args.requests {
        let permit = semaphore
            .clone()
            .acquire_owned()
            .await
            .expect("acquire concurrency permit");
        let client = client.clone();
        let url = args.url.clone();
        tasks.push(tokio::spawn(async move {
            let _permit = permit;
            request_once(&client, &url).await
        }));
    }

    let mut samples = Vec::with_capacity(args.requests);
    for task in tasks {
        samples.push(task.await.expect("join request task"));
    }

    print_summary(&args, started.elapsed(), &samples);
}

async fn request_once(client: &Client, url: &str) -> ResultSample {
    let body = json!({
        "model": "claude-load-test",
        "max_tokens": 16,
        "messages": [{"role": "user", "content": "ping"}],
    });
    let started = Instant::now();
    let status = match client
        .post(url)
        .header("content-type", "application/json")
        .header("authorization", "Bearer load-test")
        .json(&body)
        .send()
        .await
    {
        Ok(resp) => {
            let status = resp.status().as_u16();
            let _ = resp.bytes().await;
            status
        }
        Err(_) => 0,
    };

    ResultSample {
        status,
        latency_ms: started.elapsed().as_secs_f64() * 1000.0,
    }
}

fn print_summary(args: &Args, elapsed: Duration, samples: &[ResultSample]) {
    let mut statuses = BTreeMap::new();
    let mut latencies = samples
        .iter()
        .map(|sample| sample.latency_ms)
        .collect::<Vec<_>>();
    latencies.sort_by(|a, b| a.total_cmp(b));

    for sample in samples {
        *statuses.entry(sample.status).or_insert(0usize) += 1;
    }

    let ok = statuses.get(&200).copied().unwrap_or(0);
    let errors = args.requests.saturating_sub(ok);
    let elapsed_secs = elapsed.as_secs_f64();
    println!("{{");
    println!("  \"requests\": {},", args.requests);
    println!("  \"concurrency\": {},", args.concurrency);
    println!("  \"elapsed_seconds\": {:.3},", elapsed_secs);
    println!(
        "  \"requests_per_second\": {:.2},",
        args.requests as f64 / elapsed_secs
    );
    println!("  \"ok\": {ok},");
    println!("  \"errors\": {errors},");
    println!("  \"statuses\": {:?},", statuses);
    println!("  \"latency_ms\": {{");
    println!(
        "    \"min\": {:.2},",
        latencies.first().copied().unwrap_or(0.0)
    );
    println!("    \"mean\": {:.2},", mean(&latencies));
    println!("    \"p50\": {:.2},", percentile(&latencies, 50.0));
    println!("    \"p95\": {:.2},", percentile(&latencies, 95.0));
    println!("    \"p99\": {:.2},", percentile(&latencies, 99.0));
    println!(
        "    \"max\": {:.2}",
        latencies.last().copied().unwrap_or(0.0)
    );
    println!("  }}");
    println!("}}");
}

fn mean(values: &[f64]) -> f64 {
    if values.is_empty() {
        return 0.0;
    }
    values.iter().sum::<f64>() / values.len() as f64
}

fn percentile(values: &[f64], pct: f64) -> f64 {
    if values.is_empty() {
        return 0.0;
    }
    let index = ((pct / 100.0) * (values.len().saturating_sub(1)) as f64).round() as usize;
    values[index.min(values.len() - 1)]
}

fn parse_args() -> Args {
    let mut args = Args {
        url: "http://127.0.0.1:3000/v1/messages".into(),
        requests: 10_000,
        concurrency: 128,
        timeout_secs: 10,
    };
    let mut iter = std::env::args().skip(1);

    while let Some(arg) = iter.next() {
        match arg.as_str() {
            "--url" => args.url = iter.next().expect("--url value"),
            "--requests" => {
                args.requests = iter
                    .next()
                    .expect("--requests value")
                    .parse()
                    .expect("valid --requests")
            }
            "--concurrency" => {
                args.concurrency = iter
                    .next()
                    .expect("--concurrency value")
                    .parse()
                    .expect("valid --concurrency")
            }
            "--timeout" => {
                args.timeout_secs = iter
                    .next()
                    .expect("--timeout value")
                    .parse()
                    .expect("valid --timeout")
            }
            "--help" | "-h" => {
                print_help();
                std::process::exit(0);
            }
            other => panic!("unknown argument {other}"),
        }
    }

    args
}

fn print_help() {
    println!("Usage: load_test [--url URL] [--requests N] [--concurrency N] [--timeout SECS]");
}
